daemon/api.go

```diff
@@ -2352,6 +2360,80 @@ func postCreateUser(c *Command, r *http.Request, user *auth.UserState) Response
 	}, nil)
 }
 
+type postUserRemoveData struct {
+	Username string `json:"username"`
+}
+
+func postRemoveUser(c *Command, r *http.Request, user *auth.UserState) Response {
+	_, uid, _, err := postRemoveUserUcrednetGet(r.RemoteAddr)
+	if err != nil {
+		return BadRequest("cannot get ucrednet uid: %v", err)
+	}
+	if uid != 0 {
+		return BadRequest("cannot use remove-user as non-root")
+	}
+
+	var removeData postUserRemoveData
+
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&removeData); err != nil {
+		return BadRequest("cannot decode remove-user data from request body: %v", err)
+	}
+
+	// verify request
+	st := c.d.overlord.State()
+	st.Lock()
+	users, err := auth.Users(st)
+	st.Unlock()
+	if err != nil {
+		return InternalError("cannot get user count: %s", err)
+	}
+
+	username := removeData.Username
+	if username == "" {
+		return BadRequest("cannot remove user: 'username' field is empty")
+	}
+
+	// check if user can be removed
+	if _, err := userLookup(username); err != nil {
+		return BadRequest("cannot remove user %s: %s", username, err)
+	}
+	id, err := getUserID(username, users)
+	if err != nil {
+		return BadRequest("cannot remove user %s: user is a system user", username)
+	}
+
+	// TODO: check options
+	opts := &osutil.UserDelOptions{
+		Force:     false,
+		Remove:    true,
+		ExtraUser: !release.OnClassic,
+	}
+
+	// Remove user from system
+	if err := osutilUserDel(username, opts); err != nil {
+		return BadRequest("cannot remove user %s: %s", username, err)
```

Comment:
because we don't keep a lock ideally here we should the ignore the error if the user doesn't exist, either by we understanding the tool error or we lookup the user again